### PR TITLE
Openstack provider: kube-up implementation

### PR DESCRIPTION
**Pull request by zreigz**

### CODE REVIEW

**Comment by zreigz** [comment]
@ohtake can you review this provider ?

File: cluster/openstack/config-default.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+# Flag indicates if new image must be created. If 'false' then image with IMAGE_ID will be used.
+# If 'true' then new image will be created from file config-image.sh
+CREATE_IMAGE="true" # use "true" for devstack
+
+# Image id which will be used for kubernetes stack
+IMAGE_ID="f0f394b1-5546-4b68-b2bc-8abe8a7e6b8b"
+
+# DNS server address
+DNS_SERVER="172.24.38.123"
```
Comment (cheld) [inline cluster/openstack/config-default.sh]: What is that?

File: cluster/openstack/config-default.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+MAX_NUMBER_OF_MINIONS="1"
+
+MASTER_FLAVOR="m1.small"
+
+MINION_FLAVOR="m1.small"
+
+EXTERNAL_NETWORK="public"
+
+SWIFT_SERVER_URL="http://192.168.123.100:8080"
```
Comment (cheld) [inline cluster/openstack/config-default.sh]: So, you set the defaults for DevStack?

File: cluster/openstack/config-image.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# Image name which will be displayed in OpenStack
+OPENSTACK_IMAGE_NAME="CentOS7"
+
+# Downloaded image name for Openstack project
+IMAGE_FILE="CentOS-7-x86_64-GenericCloud-1510.qcow2"
+
+# Absolute path where image file is stored.
+IMAGE_PATH="/home/lukasz/openstack_temp"
+
```
Comment (cheld) [inline cluster/openstack/config-image.sh]: Please use public URL. Left over, I guess

File: cluster/openstack/openrc-swift.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+
+# Copyright 2015 The Kubernetes Authors All rights reserved.
```
Comment (cheld) [inline cluster/openstack/openrc-swift.sh]: This file does not make sense to check in, does it?

File: cluster/openstack/openrc-default.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
```
Comment (cheld) [inline cluster/openstack/openrc-default.sh]: should not be checked in, I assume

**Comment by ohtake** [comment]
Could you split one huge commit into logically-separated commits?
There should be at least two commits.
If logical, more than three is better.

1. The original heat in our GitLab. Because the original heat can be executed independently.
2. Shell scripts called from kube-up.sh and some modification on heat if any

File: cluster/openstack/util.sh [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  fi
+}
+
+# Create a new glance image.
+#
+# Assumed vars:
+#   IMAGE_FILE
+#   IMAGE_PATH
+#   OPENSTACK_IMAGE_NAME
+function create-glance-image() {
```
Comment (cheld) [inline cluster/openstack/util.sh]: We do not deliver the image. It must is provided by the cloud provider or the user. The other provider just assume that the required base image is existing. Maybe we should create a separate script utility script to upload CentOS7?

File: cluster/openstack/kubernetes-heat/kubecluster.yaml [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        - protocol: tcp # etcd for flanneld
+          port_range_min: 4379
+          port_range_max: 4380
+        - protocol: udp # flannel backend
+          port_range_min: 8285
+          port_range_max: 8285
+        - protocol: tcp # api-server
+          port_range_min: 443
+          port_range_max: 443
+        - protocol: tcp
```
Comment (ohtake) [inline cluster/openstack/kubernetes-heat/kubecluster.yaml]: These are my corner-cutting. I left port listings (7080, 8080, 2379, 6443) in the old heat-provider and added required ports (4379-4380, 8285, 443) in salt-based provisioning. Ports left behind should be removed if unnecessary.

File: cluster/openstack/util.sh (hunk 6 above)
Comment (zreigz) [inline cluster/openstack/util.sh]: There is a flag in config-default.sh called CREATE_IMAGE. This function is useful when we use devstack. 
If you want use image from provider just setup:
IMAGE_ID="f0f394b1-5546-4b68-b2bc-8abe8a7e6b8b"
and CREATE_IMAGE=flase

**Comment by cheld** [comment]
@ohtake 

> Could you split one huge commit into logically-separated commits?

I agree on smaller commits, but now its already too late I guess. 

> The original heat in our GitLab. Because the original heat can be executed independently.

in best case we would like to contribute this code. For a few reasons, e.g. lack of K5 support, this will not happen soon. But never the less we should continue in this direction.

I don't see any reason we would want to continue using the old heat repository in gitlab. It would just make our life more complicated. Could you explain you reasoning?

File: cluster/openstack/config-default.sh (hunk 2 above)
Comment (zreigz) [inline cluster/openstack/config-default.sh]: yes

File: cluster/openstack/kubernetes-heat/kubecluster.yaml (hunk 7 above)
Comment (cheld) [inline cluster/openstack/kubernetes-heat/kubecluster.yaml]: We should not change the salt scripts, because we cannot contribute any longer

**Comment by ohtake** [comment]
I don't intend to use the old repository. This provider consists of two layers. The former one is heat, and latter is wrapper for kube-up.sh. They can be splitted logically.

File: cluster/openstack/kubernetes-heat/kubecluster.yaml (hunk 7 above)
Comment (ohtake) [inline cluster/openstack/kubernetes-heat/kubecluster.yaml]: No need to change salt scripts.
Ports 7080, 8080, 2379, 6443 are used by [larsks/heat-kubernetes](https://github.com/larsks/heat-kubernetes).
I am not sure if these ports are used by a salt-based cluster.
I just left them as they were, and add ports 4379-4380, 8285, 443 which are surely necessary for a salt-based cluster.
I should have investigated whether the old ports were required or not.

File: cluster/openstack/config-default.sh (hunk 1 above)
Comment (zreigz) [inline cluster/openstack/config-default.sh]: I will change it for public DNS like 8.8.8.8

**Comment by cheld** [comment]
@ohtake ok I see

File: cluster/openstack/config-image.sh (hunk 3 above)
Comment (zreigz) [inline cluster/openstack/config-image.sh]: I wanted make it more generic to be able use different images. Public URL contains archived image. So it is better have local storage for images

File: cluster/openstack/kubernetes-heat/fragments/provision-network-node.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+fi
+
+cat <<EOF >/etc/sysconfig/flanneld
+FLANNEL_ETCD="${FLANNEL_ETCD_URL}"
+FLANNEL_ETCD_KEY="/coreos.com/network"
+FLANNEL_OPTIONS="-iface=eth0 --ip-masq"
+EOF
+
+systemctl enable flanneld
+systemctl restart flanneld
```
Comment (ohtake) [inline cluster/openstack/kubernetes-heat/fragments/provision-network-node.sh]: Please cherry-pick a commit from GitLab. If a user deploys a cluster on DevStack, some additional configuration is required.

**Comment by zreigz** [comment]
yes but it still one part as provider. We want it together. Two commits won't work separately. For me this is one logical part called provider, share one directory.

File: cluster/openstack/kubernetes-heat/fragments/provision-network-node.sh (hunk 8 above)
Comment (zreigz) [inline cluster/openstack/kubernetes-heat/fragments/provision-network-node.sh]: For this stage I can copy and paste to this provider structure. @ohtake  is it ok for you

Comment (zreigz) [inline cluster/openstack/kubernetes-heat/fragments/provision-network-node.sh]: I've added missing part

**Comment by zreigz** [comment]
PTAL

**Comment by cheld** [comment]
Btw: can you fix travis? I think some header wrong or missing

File: cluster/openstack/openrc-default.sh (hunk 5 above)
Comment (zreigz) [inline cluster/openstack/openrc-default.sh]: Those files are needed for Openstack clients like swift ,nova heat. Without those files you can not authenticate your clients

**Comment by cheld** [comment]
Can I merge?

File: cluster/openstack/openrc-default.sh (hunk 5 above)
Comment (cheld) [inline cluster/openstack/openrc-default.sh]: But they contain tenant specific data, right? They might work for you but not for anyone else, I assume.

**Comment by zreigz** [comment]
LGTM You can merge.

**Pull request merged**